src/main.rs

```diff
@@ -33,43 +51,122 @@ fn init_logger() {
+    // is guaranteed to be a valid filepath, which is checked by
+    // JobConfig::from_file in main.rs
+    let sources_dir = opt
+        .jobconfig
+        .parent()
+        .expect("Invalid jobconfig path")
+        .to_owned();
+    let noclean = opt.noclean;
 
-    let deploy_prefix;
-    if let Some(sources) = config.sources {
-        // It's safe to call unwrap here - at this point opt.jobconfig
-        // is guaranteed to be a valid filepath, which is checked by
-        // JobConfig::from_file
-        let prefix = mgr
-            .deploy(opt.jobconfig.parent().unwrap(), &sources, &config.progname)
-            .context("deploying the sources")?;
-        deploy_prefix = Some(prefix);
-    } else {
-        deploy_prefix = None;
-    }
+    SessionMPI::init(opt.hub, prov_filter)
+        .handle_ctrlc()
+        .context("initializing session")
+        .and_then(move |session| {
+            use std::rc::Rc;
+            let session = Rc::new(session);
+            let mut session_clone = Rc::clone(&session);
 
-    mgr.exec(
-        opt.numproc,
-        config.progname,
-        config.args,
-        config.mpiargs,
-        deploy_prefix,
-    )?;
+            info!("available cores: {}", session.total_cpus());
+            let cpus_available = session.total_cpus();
+            if cpus_available < cpus_requested {
+                return Either::A(future::err(format_err!(
+                    "Not enough CPUs available: requested: {}, available: {}",
+                    cpus_requested,
+                    cpus_available
+                )));
+            }
+            info!("Compiling the sources...");
+            let deploy_prefix = if let Some(sources) = config.sources.clone() {
+                Either::A(
+                    session
+                        .deploy(sources_dir, sources, progname)
+                        .context("deploying the sources")
+                        .and_then(|depl| {
+                            let DeploymentInfo {
+                                logs,
+                                deploy_prefix,
+                            } = depl;
+                            for comp in logs {
+                                let logs = comp.logs.join("\n------------------\n");
+                                info!(
+                                    "Provider {} compilation output:\n{}",
+                                    comp.node.to_string(),
+                                    logs
+                                );
+                            }
+                            Ok(Some(deploy_prefix))
+                        }),
+                )
+            } else {
+                Either::B(future::ok(None))
+            };
 
-    if let Some(output) = config.output {
-        mgr.retrieve_output(&output)?;
-    }
+            Either::B(
+                deploy_prefix
+                    .and_then(move |deploy_prefix| {
```

Comment:
Please consider moving those closures into separate named functions to improve readability. I find it easier to grasp code at first when I see only things like 'deploying, compiling, retrieve_output, cleanup. What do you think?

Reply:
This will be greatly simplified with the advent of async/await. Currently we need a lot of boilerplate to connect the functions from SessionMPI together.